### PR TITLE
EthStorage Mining 

### PR DESCRIPTION
**Pull request by ping-ke**
related design is added to the following doc: 
https://docs.google.com/document/d/1ePI2Bqj069ffKV9eTYbRHCtX5ljyxH3P1BKtU3SGgWU/edit#

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2599,3 +2645,55 @@ func (bc *BlockChain) GetSstorageLastKvIdx(contract common.Address) (uint64, err
...
+	if err != nil {
+		return nil, err
+	}
+
+	return bc.GetSstorageMiningInfoWithStateDB(stateDB, contract, shardId)
+}
+
+func (bc *BlockChain) GetSstorageMiningInfoWithStateDB(stateDB *state.StateDB, contract common.Address, shardId uint64) (*MiningInfo, error) {
+	info := new(MiningInfo)
+	position := getSlotHash(0, uint256.NewInt(shardId).Bytes32())
```
Comment (qzhodl) [inline core/blockchain.go:2686]: DecentralizedKVDaggerHashimoto inherits from DecentralizedKV, so the info mapping is not at slot 0

File: sstorminer/worker.go [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+			kvs, err := w.chain.ReadKVsByIndexList(t.storageContract, kvIdxs, true)
+			if err != nil {
+				return false, err
+			}
+			if len(kvs) != len(kvIdxs) {
+				return false, fmt.Errorf("fail to get all the kvs %v", kvIdxs)
+			}
+
+			for i := 0; i < len(dataSet); i++ {
+				if kvs[i].Idx != kvIdxs[i] {
```
Comment (qzhodl) [inline sstorminer/worker.go:674]: should be == instead of !=

File: sstorminer/worker.go [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+		Value:     new(big.Int).SetInt64(0),
+		Data:      data,
+	}
+
+	signedTx, err := types.SignTx(types.NewTx(baseTx), w.signer, w.privKey)
+	if err != nil {
+		return err
+	}
+
+	return w.eth.TxPool().AddLocal(signedTx)
```
Comment (qzhodl) [inline sstorminer/worker.go:730]: Will AddLocal trigger to broadcast the tx?

Comment (ping-ke) [reply sstorminer/worker.go:719]: yes, it can

File: eth/protocols/sstorage/sync.go [hunk 4]
```diff
@@ -492,6 +492,19 @@ func (s *Syncer) loadSyncStatus() {
 			lastKvIndex = 0
 		}
 		for _, sid := range shards {
+			// if the shard is not full, fill in the data shard with []byte{},
```
Comment (qizhou) [inline eth/protocols/sstorage/sync.go:495]: I don't think it is a good idea here:

1.  It seems that this will create a race condition that sm.TryWrite() may override the data that are synced by other threads.
2.  The single-threaded write will take a considerably long time to finish.

Why not write the empty data when issuing requests and check if the request index > LastKvIndex, then we will just write empty data?

File: sstorminer/worker.go [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+		taskDoneCh:         make(chan struct{}),
+		taskStartCh:        make(chan struct{}),
+		signer:             types.NewEIP2930Signer(chainConfig.ChainID),
+		privKey:            privKey,
+	}
+	for addr, sm := range sstor.ContractToShardManager {
+		for idx, shard := range sm.ShardMap() {
+			task := task{
+				storageContract: addr,
+				minerContract:   minerContract,
```
Comment (qzhodl) [inline sstorminer/worker.go:300]: do we want to separate the miner contract and storage contract?

Comment (ping-ke) [reply sstorminer/worker.go:300]: currently, yes. and the mining contract may need to update to support this.

File: sstorminer/worker.go [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+		select {
+		case <-w.startCh:
+			w.updateTaskInfo(w.chain.CurrentBlock().Root(), time.Now().Unix())
+			timer.Reset(recommit)
+
+		case head := <-w.chainHeadCh:
+			w.updateTaskInfo(head.Block.Root(), time.Now().Unix())
+			timer.Reset(recommit)
+
+		case <-timer.C:
```
Comment (qzhodl) [inline sstorminer/worker.go:419]: chain header change has already reset the timer for 3s intervals (for our testnet case)

File: sstorminer/worker.go [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+func (w *worker) taskLoop() {
+	defer w.wg.Done()
+
+	for {
+		select {
+		case task := <-w.taskCh:
+			if w.newTaskHook != nil {
+				w.newTaskHook(task)
+			}
+			_, err := w.mineTask(task)
```
Comment (qzhodl) [inline sstorminer/worker.go:493]: do we want to enable parallel mining for all the shards?

File: sstorminer/worker.go [hunk 8]
```diff
@@ -0,0 +1,728 @@
...
+	worker := &worker{
+		config:             config,
+		chainConfig:        chainConfig,
+		eth:                eth,
+		mux:                mux,
+		chain:              chain,
+		tasks:              make([]*task, 0),
+		chainHeadCh:        make(chan core.ChainHeadEvent, chainHeadChanSize),
+		newWorkCh:          make(chan *newWorkReq),
+		taskCh:             make(chan *task, len(sstor.Shards())),
```
Comment (qzhodl) [inline sstorminer/worker.go:286]: discussed with Ping, the channel buffer should be 0 in the current design

Comment (qzhodl) [inline sstorminer/worker.go:286]: discussed with Ping, the channel buffer should be 0 in the current design

File: eth/protocols/sstorage/sync.go (hunk 4 above)
Comment (ping-ke) [reply eth/protocols/sstorage/sync.go]: update

File: sstorminer/worker.go (hunk 8 above)
Comment (ping-ke) [reply sstorminer/worker.go:286]: updated

File: sstorminer/worker.go (hunk 7 above)
Comment (ping-ke) [reply sstorminer/worker.go:493]: will support this later

File: core/blockchain.go [hunk 9]
```diff
@@ -2448,6 +2455,34 @@ func VerifyKV(sm *sstorage.ShardManager, idx uint64, val []byte, meta *SstorageM
...
+	empty := make([]byte, 0)
+	lastKvIdx, err := bc.GetSstorageLastKvIdx(contract)
+	if err != nil {
+		return start, fmt.Errorf("get lastKvIdx for FillEmptyKV fail, err: %s", err.Error())
+	}
+	for idx := start; idx <= limit; idx++ {
+		if lastKvIdx > idx {
+			continue
+		}
+		_, err = sm.TryWrite(idx, empty, common.Hash{})
```
Comment (qzhodl) [inline core/blockchain.go:2476]: The hash of zero data block should not be zero

Comment (ping-ke) [reply core/blockchain.go:2476]: that is because merkleRootWithMinTree returns common.Hash{} for empty data.

File: sstorminer/worker.go (hunk 6 above)
Comment (ping-ke) [reply sstorminer/worker.go:424]: Timer use to refresh for edge cases and tests.

**Comment by ping-ke** [comment]
This code has been reviewed,  and deployed to test https://testnet.web3q.io:8545/, and do some integration tests. There are still 2 bugs remaining, trace them as follow: [put func of system contract 03330001 do not return correct data.](https://github.com/QuarkChain/go-ethereum/issues/140), [ethstorage mine tranction will fail with "execution reverted"](https://github.com/QuarkChain/go-ethereum/issues/140)

File: eth/sync.go [hunk 10]
```diff
@@ -31,7 +31,7 @@ import (
 
 const (
 	forceSyncCycle      = 10 * time.Second // Time interval to force syncs, even if few peers are available
-	defaultMinSyncPeers = 5                // Amount of peers desired to start syncing
+	defaultMinSyncPeers = 2                // Amount of peers desired to start syncing
```
[the remark this answers is not in the document]
Comment (ping-ke) [reply eth/sync.go:34]: this param is used to control when to start sync, as our test net node is less, I change this value to a small one.

	minPeers := defaultMinSyncPeers
	if cs.forced {
		minPeers = 1
	} else if minPeers > cs.handler.maxPeers {
		minPeers = cs.handler.maxPeers
	}
	if cs.handler.peers.len() < minPeers {
		return nil
	}

File: core/genesis.go [hunk 11]
```diff
@@ -446,7 +446,7 @@ func DefaultWeb3QTestnetGenesisBlock() *Genesis {
 			common.HexToAddress("0x977cfc676bb06daed7ddfa7711bcfe8d50c93081"),
 			common.HexToAddress("0xcd21538af6e33ff6fcf1e2ca20f771413004cfd3"),
 		},
-		NextValidatorPowers: []uint64{1, 1, 1, 1},
+		NextValidatorPowers: []uint64{11, 11, 11, 11},
```
[the remark this answers is not in the document]
Comment (ping-ke) [reply core/genesis.go:449]: remove this change

File: consensus/tendermint/gov/gov.go [hunk 12]
```diff
@@ -141,7 +141,7 @@ func (g *Governance) NextValidatorsAndPowersAt(remoteChainNumber uint64, hash co
 }
 
 // getValidatorsAndPowersFromContract get next validators from contract
-func (g *Governance) getValidatorsAndPowersFromContract(blockHash common.Hash) ([]common.Address, []uint64, error) {
```
[the remark this answers is not in the document]
Comment (ping-ke) [reply consensus/tendermint/gov/gov.go:144]: as the Ethereum header change, our etherclient cannot get the correct hash, so when we get validators and powers from the contract from Ethereum, it will fail. so change the getValidatorsAndPowersFromContract using a hash to block number.

[the remark this answers is not in the document]
Comment (ping-ke) [reply consensus/tendermint/gov/gov.go:144]: can move out this change and create a issue to check this issue.
